test/organisations_api_test.rb

```diff
@@ -4,54 +4,216 @@
+              organisation_matchers(agriculture),
+            ],
+          },
+          headers: {
+            "Content-Type" => "application/json; charset=utf-8",
+          },
+        )
+    end
+
     it "should get the organisations" do
-      organisation_slugs = %w[ministry-of-fun tea-agency]
-      stub_organisations_api_has_organisations(organisation_slugs)
+      response = mocked_api.organisations
+      first_id = "#{base_api_url}/api/organisations/#{education}"
+      assert_equal first_id, response["results"].first["id"]
+      assert_equal 200, response.code
+    end
+  end
+
+  describe "pagination" do
+    let(:pact_organisation_endpoint) { "http://example.org/api/organisations" }
+    let(:page_one_links) { %(<#{pact_organisation_endpoint}?page=2>; rel="next", <#{pact_organisation_endpoint}?page=1>; rel="self") }
+    let(:page_two_links) { %(<#{pact_organisation_endpoint}?page=1>; rel="previous", <#{pact_organisation_endpoint}?page=2>; rel="self") }
 
-      response = @api.organisations
-      assert_equal(organisation_slugs, response.map { |r| r["details"]["slug"] })
-      assert_equal "Tea Agency", response["results"][1]["title"]
+    let(:request) do
+      {
+        method: :get,
+        path: "/api/organisations",
+        headers: GdsApi::JsonClient.default_request_headers,
+      }
+    end
+    let(:body) do
+      {
+        results: Pact.each_like({}, min: 20),
+        page_size: 20,
+        pages: 2,
+      }
+    end
+    let(:response) do
+      {
+        status: 200,
+        body: body,
+      }
     end
 
-    it "should handle the pagination" do
-      organisation_slugs = (1..50).map { |n| "organisation-#{n}" }
-      stub_organisations_api_has_organisations(organisation_slugs)
+    before do
+      organisation_api
+        .given("organisation results are paginated")
+        .upon_receiving("a request without a query param")
+        .with(request.merge(query: ""))
+        .will_respond_with(response.merge(headers: { "link" => page_one_links }))
 
-      response = @api.organisations
-      assert_equal(
-        organisation_slugs,
-        response.with_subsequent_pages.map { |r| r["details"]["slug"] },
-      )
+      organisation_api
+        .given("organisation results are paginated")
+        .upon_receiving("a request with page 2 params")
+        .with(request.merge(query: "page=2"))
+        .will_respond_with(response.merge(headers: { "link" => page_two_links }))
     end
 
-    it "should raise error if endpoint 404s" do
-      stub_request(:get, "#{@base_api_url}/api/organisations").to_return(status: 404)
-      assert_raises GdsApi::HTTPNotFound do
-        @api.organisations
+    it "should handle pagination" do
+      def json(param = nil)
+        GdsApi::JsonClient.new.get_raw("#{organisation_api_host}/api/organisations#{param}")
       end
+
+      # Stub the response to get_list called by GdsApi::Organisations.organisations, because the Pact tests
```

Comment:
I had a look at this, and unfortunately couldn't get to the bottom of it. For some reason in Collections `request.host` is `example.org`. It doesn't matter what the host is, and if we could force it to be `localhost:4000`, then we could remove these stubs. But I couldn't figure out where it's coming from.

(Would be great if you're able to figure it out, as I expect it'll cause pain in future tests.) I think it would be helpful to give a bit more info in this comment, as it's quite a subtle problem. I think the main points are:

- Pagination relies on "next" links in the response headers / body.
- The link includes the host, which we could set to the api host here.
- But on the provider side, it gets overridden to `example.org`.
- We therefore have to cope with `example.org` here so it works for both 😠.

Reply:
Ah ha! Figured it out. See [here](https://github.com/rails/rails/blob/291a3d2ef29a3842d1156ada7526f4ee60dd2b59/actionpack/lib/action_controller/renderer.rb#L39) and [here](https://stackoverflow.com/questions/598933/how-do-i-change-the-default-www-example-com-domain-for-testing-in-rails).

Will remove the stubs and amend in collections eg
```
RSpec.configure do |config|
  config.before(:each, type: :request) do
    host! "localhost:4000"
  end
end
```

Reply:
I've created a proxy so we can overwrite the default request.host in collections 

see: https://github.com/alphagov/collections/pull/2230/commits/946e2d3592ab2da17011064ee21bf14f7edc1db4